db/crud.go

```diff
@@ -842,6 +842,203 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string,
+			base.WarnfCtx(db.Ctx, base.KeyAll, "Unable to marshal document body for storage in rev tree: %v", marshalErr)
+		}
+		doc.setNonWinningRevisionBody(prevCurrentRev, bodyJSON, db.AllowExternalRevBodyStorage())
+	}
+	// Store the new revision body into the doc:
+	storedBody := doc.setRevisionBody(newRevID, body, db.AllowExternalRevBodyStorage())
+
+	if doc.CurrentRev == newRevID {
+		doc.NewestRev = ""
+		doc.setFlag(channels.Hidden, false)
+	} else {
+		doc.NewestRev = newRevID
+		doc.setFlag(channels.Hidden, true)
+		if doc.CurrentRev != prevCurrentRev {
+			doc.promoteNonWinningRevisionBody(doc.CurrentRev, db.RevisionBodyLoader)
+		}
+	}
+	return storedBody
+}
+
+func (db *Database) runSyncFn(doc *Document, body Body, newRevId string) (*uint32, string, base.Set, channels.AccessMap, channels.AccessMap, error) {
+	channelSet, access, roles, syncExpiry, oldBody, err := db.getChannelsAndAccess(doc, body, newRevId)
+	if err != nil {
+		return nil, ``, nil, nil, nil, err
+	}
+	db.checkDocChannelsAndGrantsLimits(doc.ID, channelSet, access, roles)
+
+	return syncExpiry, oldBody, channelSet, access, roles, nil
+}
+
+func (db *Database) recalculateSyncFnForActiveRev(doc *Document, newRevID string) (channelSet base.Set, access, roles channels.AccessMap, syncExpiry *uint32, oldBodyJSON string, err error) {
+	// In some cases an older revision might become the current one. If so, get its
+	// channels & access, for purposes of updating the doc:
+	var curBody Body
+	if curBody, err = db.getAvailableRev(doc, doc.CurrentRev); curBody != nil {
+		base.DebugfCtx(db.Ctx, base.KeyCRUD, "updateDoc(%q): Rev %q causes %q to become current again",
+			base.UD(doc.ID), newRevID, doc.CurrentRev)
+		channelSet, access, roles, syncExpiry, oldBodyJSON, err = db.getChannelsAndAccess(doc, curBody, doc.CurrentRev)
+		if err != nil {
+			return
+		}
+	} else {
+		// Shouldn't be possible (CurrentRev is a leaf so won't have been compacted)
+		base.WarnfCtx(db.Ctx, base.KeyAll, "updateDoc(%q): Rev %q missing, can't call getChannelsAndAccess "+
+			"on it (err=%v)", base.UD(doc.ID), doc.CurrentRev, err)
+		channelSet = nil
+		access = nil
+		roles = nil
+	}
+	return
+}
+
+func (db *Database) addAttachments(newAttachments AttachmentData) error {
+	// Need to check and add attachments here to ensure the attachment is within size constraints
+	err := db.setAttachments(newAttachments)
+	if err != nil {
+		if err.Error() == "document value was too large" {
+			err = base.HTTPErrorf(http.StatusRequestEntityTooLarge, "Attachment too large")
+		} else {
+			err = errors.Wrap(err, "Error adding attachment")
+		}
+	}
+	return err
+}
+
+// Sequence processing :
+// Assigns provided sequence to the document
+// Update unusedSequences in the event that there is a conflict and we have to provide a new sequence number
+// Update and prune RecentSequences
+func (db *Database) handleSequences(docSequence uint64, doc *Document, unusedSequences []uint64) ([]uint64, error) {
```

Comment:
I think renaming this to 'assignSequence' would make the functionality clearer.  The unused/recent sequence work is really just follow-on from assigning the sequence.